### DISCUSSION TITLE
Control diag_kind/kind individually in pairplot

### DISCUSSION THREAD

**Discussion by lwelzel** [mwaskom/seaborn#3189] [Q&A]
I would like to have control over the kind of plotting function that gets called for the subplots in pairplot:

For some applications the data supplied to pairplot can be mixed continues, ordinal and/or nominal/discrete (think e.g. hyper-parameters). This means that plotting each individual variable pair would benefit from a separate choice between scatter/kde/hist, especially for the diagonal.

It seems this is currently not supported using the kind, diag_kind or {plot, diag, grid}_kws parameters.
- For a N variable pairplot diag_kind should accept a sequence/list of length N with the plot style.
 - For a N variable pairplot kind should accept either sequence/list of length N^2 (or N! for a triangle) with each individual plot style or of length N where it defaults to the default when there is a conflict.

For the example given in the pairplot api, this could look like this:
```
# bill_length, bill_depth, flipper_length, body_mass
diag_kind = [‘hist’, ‘kde’, ‘hist’, ‘kde’]
# or: 
kind = [[‘hist’, ‘kde’, ‘scatter’, ‘hist’], 
[‘kde’, ‘kde’, ‘scatter’, ‘scatter’], 
[‘scatter’, ‘scatter’, ‘hist’, ‘scatter’], 
[‘hist’, ‘scatter’, ‘scatter’, ‘kde’]]
```
with the output:
<img width="563" alt="pairplot_ind_control" src="https://user-images.githubusercontent.com/29613344/206724106-cfcd4fbc-9ec9-4389-a5d0-dd9e2b17d802.png">

I am not very familiar with the code base but this seems like it could be accomplished by calling the appropriate function for each subplot based on the kind/diag_kind keyword without major changes.

Edit: I did not check if this is already possible using the returned PairGrid.

**Comment by mwaskom** [comment]
This is not supported in `PairGrid` and hence isn't just a simple addition to `pairplot`.

I am sympathetic to the idea of wanting to mix numeric / categorical variables into one plot and to use different representations for them but I am skeptical of the value of a fully configurable "kind for each subplot" API.

I'm not sure how you made the plot in your post but there are a variety of ways to leverage existing functionality to get more flexibility, e.g.

```python
def flexplot(x, y, **kwargs):
    if "flipper_length_mm" in (x.name, y.name):
        func = sns.kdeplot
    else:
        func = sns.scatterplot
    func(x=x, y=y, **kwargs)

g = sns.PairGrid(penguins)
g.map_offdiag(flexplot)
g.map_diag(sns.histplot)
```
![image](https://user-images.githubusercontent.com/315810/208278913-daaff6f6-5c86-42ff-829f-7db8470ceaee.png)

My main objection to your proposal is that writing down that matrix of kinds is quite cumbersome, and if you're at the point where you're doing that with logic, it's only marginally harder to integrate the plotting into the logic itself. There's also a notion of scope creep; e.g. how do you generalize `plot_kws` to pass different parameters to different function types? Ultimately, functions like `pairplot` work because there are simple rules; trying to stuff a more flexible API into a high-level interface is misguided and it's better to allow people to make more bespoke plots with the next level down of abstraction.
